### PR TITLE
fix: Apply Tags consistently across NIC, VM and Disks

### LINKED ISSUE #75: Tag network interfaces with `karpenter.sh_nodepool`
### Tell us about your request

all resources created by the Karpenter provider should have consistent tags in Azure. For example: `karpenter.azure.com_cluster` & `karpenter.sh_nodepool`.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

Tag und Label consistency for calls like `az resource list --tag=karpenter.sh_nodepool` or `az resource list --tag=karpenter.azure.com_cluster`. At the moment disks & vms get both tags assigned but nics only have the tag `karpenter.azure.com_cluster`.

### Are you currently working around this issue?

Only working with the label / tag `karpenter.azure.com_cluster`

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by iamvighnesh**
Fixes #75 

**Description**

This fix consolidates Tags applied on various Azure resources like NIC, VMs and Disks in one place.

**How was this change tested?**

* Tested manually with multiple Karpenter managed node pools
* Acceptance tests are also added for coverage

**Does this change impact docs?**
- [ ] No

**Release Note**

NONE

### CODE REVIEW

**Review by Bryce-Soghigian — CHANGES_REQUESTED** [review]
Can you add some tests with our framework to validate the tags are set on the network interface? The tests for nic are in pkg/providers/instancetype/suite_test.go

File: pkg/providers/instance/instance_test.go [hunk 1]
```diff
@@ -149,6 +150,48 @@ func TestGetZone(t *testing.T) {
...
+					},
+				},
+			},
+			tags:         map[string]*string{},
+			expectedTags: map[string]*string{},
+		},
+	}
+
+	for _, c := range tc {
+		setNodePoolNameTag(c.tags, c.nodeClaim)
```
Comment (Bryce-Soghigian) [inline pkg/providers/instance/instance_test.go:190]: We should get coverage fine from the acceptance tests you added imo we don't need this :)

Comment (iamvighnesh) [reply pkg/providers/instance/instance_test.go]: The unit test for the function is now removed.

**Review by Bryce-Soghigian — COMMENTED** [review]
/approve cc: @Azure/karpenter  can we get another reviewer to approve to get this in? 

Note we need two approvers for external contributors

**Review by Bryce-Soghigian — APPROVED** [review]

File: pkg/operator/options/suite_test.go
Comment (tallaxes) [inline pkg/operator/options/suite_test.go:1]: These changes are not related to tagging ... Did something get in by mistake during merging?

Comment (tallaxes) [reply pkg/operator/options/suite_test.go:1]: Other that this, LGTM

Comment (iamvighnesh) [reply pkg/operator/options/suite_test.go:1]: @tallaxes these changes were added when code coverage build failed for missing coverage on those scenarios. I have added them to add coverage for the functionality and pass the coverage build too.

Comment (tallaxes) [reply pkg/operator/options/suite_test.go:1]: You added unit tests for unrelated code, to satisfy the code coverage gate? 😯 Appreciate the extra coverage, but let's do this in a separate PR next time. Note that while coverage check is a useful signal, it is not blocking.

**Review by tallaxes — COMMENTED** [review]
/test

**Review by tallaxes — COMMENTED** [review]
/test

**Pull request merged**